## Data Cleaning

* Este notebook es la primera parte del proyecto de visualización para el bootcamp IronHack - semana 5
* Se trata de una base de datos descargada de kaggle del siguiente link: https://www.kaggle.com/code/marwandiab/boat-sales-analysis

In [1]:
import pandas as pd

import sys
sys.path.append('../')
import src.support as sp

In [2]:
data = pd.read_csv('../data/boat_data.csv')
data.head()

Price            Boat Type          Manufacturer                 Type  \
0   CHF 3337          Motor Yacht  Rigiflex power boats  new boat from stock   
1   EUR 3490  Center console boat     Terhi power boats  new boat from stock   
2   CHF 3770           Sport Boat    Marine power boats  new boat from stock   
3  DKK 25900           Sport Boat    Pioner power boats  new boat from stock   
4   EUR 3399         Fishing Boat    Linder power boats  new boat from stock   

   Year Built  Length  Width       Material  \
0        2017    4.00   1.90            NaN   
1        2020    4.00   1.50  Thermoplastic   
2           0    3.69   1.42      Aluminium   
3        2020    3.00   1.00            NaN   
4        2019    3.55   1.46      Aluminium   

                                     Location  Number of views last 7 days  
0      Switzerland » Lake Geneva » Vésenaz                          226  
1                    Germany » Bönningstedt                           75  
2  Switzerland » Lake of Zurich » Stäfa ZH                          124  
3                        Denmark » Svendborg                           64  
4               Germany » Bayern » München                           58

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9888 entries, 0 to 9887
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Price                        9888 non-null   object 
 1   Boat Type                    9888 non-null   object 
 2   Manufacturer                 8550 non-null   object 
 3   Type                         9882 non-null   object 
 4   Year Built                   9888 non-null   int64  
 5   Length                       9879 non-null   float64
 6   Width                        9832 non-null   float64
 7   Material                     8139 non-null   object 
 8   Location                     9852 non-null   object 
 9   Number of views last 7 days  9888 non-null   int64  
dtypes: float64(2), int64(2), object(6)
memory usage: 772.6+ KB


Tenemos 9888 entradas, podemos ver en el .info() que hay datos nulos, sobre todo en la columna de 'Manufacturer' 

Se va a proceder a limpiar la muestra de valores nulos para poder trabajar con una visualización clara y limpia.

También se va a separar la columna 'Price' para separar en dos: valor numérico y moneda.

In [4]:
data.describe().T

count         mean         std    min      25%  \
Year Built                   9888.0  1893.192860  460.201582   0.00  1996.00   
Length                       9879.0    11.570017    6.002820   1.04     7.47   
Width                        9832.0     3.520124    1.220534   0.01     2.54   
Number of views last 7 days  9888.0   149.160801  151.819752  13.00    70.00   

                                 50%      75%      max  
Year Built                   2007.00  2017.00  2021.00  
Length                         10.28    13.93   100.00  
Width                           3.33     4.25    25.16  
Number of views last 7 days   108.00   172.00  3263.00

In [5]:
# Llamamos a la función y separamos en dos columnas aplicando función a toda la columna
data[['Precio', 'Currency']] = data['Price'].apply(sp.separa_valor_monedas).apply(pd.Series)

In [6]:
data.head()

Price            Boat Type          Manufacturer                 Type  \
0   CHF 3337          Motor Yacht  Rigiflex power boats  new boat from stock   
1   EUR 3490  Center console boat     Terhi power boats  new boat from stock   
2   CHF 3770           Sport Boat    Marine power boats  new boat from stock   
3  DKK 25900           Sport Boat    Pioner power boats  new boat from stock   
4   EUR 3399         Fishing Boat    Linder power boats  new boat from stock   

   Year Built  Length  Width       Material  \
0        2017    4.00   1.90            NaN   
1        2020    4.00   1.50  Thermoplastic   
2           0    3.69   1.42      Aluminium   
3        2020    3.00   1.00            NaN   
4        2019    3.55   1.46      Aluminium   

                                     Location  Number of views last 7 days  \
0      Switzerland » Lake Geneva » Vésenaz                          226   
1                    Germany » Bönningstedt                           75   
2  Switzerland » Lake of Zurich » Stäfa ZH                          124   
3                        Denmark » Svendborg                           64   
4               Germany » Bayern » München                           58   

    Precio Currency  
0   3337.0      CHF  
1   3490.0      EUR  
2   3770.0      CHF  
3  25900.0      DKK  
4   3399.0      EUR

In [7]:
data.Currency.unique()

array(['CHF', 'EUR', 'DKK', '£'], dtype=object)

In [8]:
data.Currency.value_counts()

Currency
EUR    8430
CHF     980
£      298
DKK     180
Name: count, dtype: int64

In [9]:
# Voy a limpiar todas aquellas donde aparezca £ y cambiarlo por GBP
data.Currency.replace('£','GBP', inplace=True)

In [10]:
data.Currency.value_counts()

Currency
EUR    8430
CHF     980
GBP     298
DKK     180
Name: count, dtype: int64

In [11]:
data['Precio_EUR'] = data.apply(sp.convert_currency, axis=1)
data.head()

Price            Boat Type          Manufacturer                 Type  \
0   CHF 3337          Motor Yacht  Rigiflex power boats  new boat from stock   
1   EUR 3490  Center console boat     Terhi power boats  new boat from stock   
2   CHF 3770           Sport Boat    Marine power boats  new boat from stock   
3  DKK 25900           Sport Boat    Pioner power boats  new boat from stock   
4   EUR 3399         Fishing Boat    Linder power boats  new boat from stock   

   Year Built  Length  Width       Material  \
0        2017    4.00   1.90            NaN   
1        2020    4.00   1.50  Thermoplastic   
2           0    3.69   1.42      Aluminium   
3        2020    3.00   1.00            NaN   
4        2019    3.55   1.46      Aluminium   

                                     Location  Number of views last 7 days  \
0      Switzerland » Lake Geneva » Vésenaz                          226   
1                    Germany » Bönningstedt                           75   
2  Switzerland » Lake of Zurich » Stäfa ZH                          124   
3                        Denmark » Svendborg                           64   
4               Germany » Bayern » München                           58   

    Precio Currency  Precio_EUR  
0   3337.0      CHF     3470.48  
1   3490.0      EUR     3490.00  
2   3770.0      CHF     3920.80  
3  25900.0      DKK     3367.00  
4   3399.0      EUR     3399.00

* Ya tengo los datos que necesito, puedo deshacerme de la columna 'Price' y poner en su lugar 'Precio', 'Currency' y 'Precio_EUR'

In [12]:
data = data.drop('Price', axis=1)

In [13]:
data.head()

Boat Type          Manufacturer                 Type  Year Built  \
0          Motor Yacht  Rigiflex power boats  new boat from stock        2017   
1  Center console boat     Terhi power boats  new boat from stock        2020   
2           Sport Boat    Marine power boats  new boat from stock           0   
3           Sport Boat    Pioner power boats  new boat from stock        2020   
4         Fishing Boat    Linder power boats  new boat from stock        2019   

   Length  Width       Material                                    Location  \
0    4.00   1.90            NaN      Switzerland » Lake Geneva » Vésenaz   
1    4.00   1.50  Thermoplastic                    Germany » Bönningstedt   
2    3.69   1.42      Aluminium  Switzerland » Lake of Zurich » Stäfa ZH   
3    3.00   1.00            NaN                        Denmark » Svendborg   
4    3.55   1.46      Aluminium               Germany » Bayern » München   

   Number of views last 7 days   Precio Currency  Precio_EUR  
0                          226   3337.0      CHF     3470.48  
1                           75   3490.0      EUR     3490.00  
2                          124   3770.0      CHF     3920.80  
3                           64  25900.0      DKK     3367.00  
4                           58   3399.0      EUR     3399.00

In [14]:
data.columns

Index(['Boat Type', 'Manufacturer', 'Type', 'Year Built', 'Length', 'Width',
       'Material', 'Location', 'Number of views last 7 days', 'Precio',
       'Currency', 'Precio_EUR'],
      dtype='object')

In [15]:
# Voy a reordenar las columnas como me apetece:
data = data[['Precio','Currency', 'Precio_EUR', 'Boat Type', 'Manufacturer', 'Type', 'Year Built', 'Length', 'Width',
       'Material', 'Location', 'Number of views last 7 days']]
data.head()

Precio Currency  Precio_EUR            Boat Type          Manufacturer  \
0   3337.0      CHF     3470.48          Motor Yacht  Rigiflex power boats   
1   3490.0      EUR     3490.00  Center console boat     Terhi power boats   
2   3770.0      CHF     3920.80           Sport Boat    Marine power boats   
3  25900.0      DKK     3367.00           Sport Boat    Pioner power boats   
4   3399.0      EUR     3399.00         Fishing Boat    Linder power boats   

                  Type  Year Built  Length  Width       Material  \
0  new boat from stock        2017    4.00   1.90            NaN   
1  new boat from stock        2020    4.00   1.50  Thermoplastic   
2  new boat from stock           0    3.69   1.42      Aluminium   
3  new boat from stock        2020    3.00   1.00            NaN   
4  new boat from stock        2019    3.55   1.46      Aluminium   

                                     Location  Number of views last 7 days  
0      Switzerland » Lake Geneva » Vésenaz                          226  
1                    Germany » Bönningstedt                           75  
2  Switzerland » Lake of Zurich » Stäfa ZH                          124  
3                        Denmark » Svendborg                           64  
4               Germany » Bayern » München                           58

In [16]:
# Ahora voy a cambiar los NaNs de la columna material con 'unknown'
data.Material.fillna('unknown', inplace=True)

In [17]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9888 entries, 0 to 9887
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Precio                       9888 non-null   float64
 1   Currency                     9888 non-null   object 
 2   Precio_EUR                   9888 non-null   float64
 3   Boat Type                    9888 non-null   object 
 4   Manufacturer                 8550 non-null   object 
 5   Type                         9882 non-null   object 
 6   Year Built                   9888 non-null   int64  
 7   Length                       9879 non-null   float64
 8   Width                        9832 non-null   float64
 9   Material                     9888 non-null   object 
 10  Location                     9852 non-null   object 
 11  Number of views last 7 days  9888 non-null   int64  
dtypes: float64(4), int64(2), object(6)
memory usage: 927.1+ KB


In [18]:
# Y procedo igual con manufacturer:
data.Manufacturer.fillna('unknown', inplace=True)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9888 entries, 0 to 9887
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Precio                       9888 non-null   float64
 1   Currency                     9888 non-null   object 
 2   Precio_EUR                   9888 non-null   float64
 3   Boat Type                    9888 non-null   object 
 4   Manufacturer                 9888 non-null   object 
 5   Type                         9882 non-null   object 
 6   Year Built                   9888 non-null   int64  
 7   Length                       9879 non-null   float64
 8   Width                        9832 non-null   float64
 9   Material                     9888 non-null   object 
 10  Location                     9852 non-null   object 
 11  Number of views last 7 days  9888 non-null   int64  
dtypes: float64(4), int64(2), object(6)
memory usage: 927.1+ KB


In [19]:
# El resto de valores nulos que haya los voy a borrar porque es info importante
data.dropna(inplace=True)
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9790 entries, 0 to 9887
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Precio                       9790 non-null   float64
 1   Currency                     9790 non-null   object 
 2   Precio_EUR                   9790 non-null   float64
 3   Boat Type                    9790 non-null   object 
 4   Manufacturer                 9790 non-null   object 
 5   Type                         9790 non-null   object 
 6   Year Built                   9790 non-null   int64  
 7   Length                       9790 non-null   float64
 8   Width                        9790 non-null   float64
 9   Material                     9790 non-null   object 
 10  Location                     9790 non-null   object 
 11  Number of views last 7 days  9790 non-null   int64  
dtypes: float64(4), int64(2), object(6)
memory usage: 994.3+ KB


In [43]:
# Listo, me quedo con esta base de datos. La paso a mi carpeta de data como boat_data_clean
data.to_csv('../data/boat_data_clean.csv')